contracts/v2/upgrader/V2Upgrader.sol

```diff
@@ -0,0 +1,170 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ */
+
+pragma solidity 0.6.11;
+
+import { SafeMath } from "@openzeppelin/contracts/math/SafeMath.sol";
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { Ownable } from "../../v1/Ownable.sol";
+import { FiatTokenV2 } from "../FiatTokenV2.sol";
+import { FiatTokenProxy } from "../../v1/FiatTokenProxy.sol";
+import { V2UpgraderHelper } from "./V2UpgraderHelper.sol";
+
+/**
+ * @title V2 Upgrader
+ * @notice Performs USDC v2 upgrade, and runs a basic sanity test in a single
+ * atomic transaction, rolling back if any issues are found. This may be
+ * overkill, but the peace of mind is worth the gas spent. By performing the
+ * upgrade atomically, it ensures that there is no disruption of service if the
+ * upgrade is not successful for some unforeseen circumstances.
+ * @dev Read docs/v2_upgrade.md
+ */
+contract V2Upgrader is Ownable {
+    using SafeMath for uint256;
+
+    FiatTokenProxy private _proxy;
+    FiatTokenV2 private _implementation;
+    V2UpgraderHelper private _helper;
+
+    /**
+     * @notice Constructor
+     * @param proxy             FiatTokenProxy contract
+     * @param implementation    FiatTokenV2 implementation contract
+     * @param helper            V2UpgraderHelper contract
+     */
+    constructor(
+        FiatTokenProxy proxy,
+        FiatTokenV2 implementation,
+        V2UpgraderHelper helper
+    ) public Ownable() {
+        _proxy = proxy;
+        _implementation = implementation;
+        _helper = helper;
+    }
+
+    /**
+     * @notice Upgrade, transfer proxy admin role to a given address, run a
+     * sanity test, and tear down the upgrader contract, in a single atomic
+     * transaction. It rolls back if there is an error.
+     * @param newName       New ERC20 name (e.g. "USD//C" -> "USD Coin")
+     * @param newProxyAdmin Grantee of the proxy admin role
+     */
+    function upgrade(string calldata newName, address newProxyAdmin)
+        external
+        onlyOwner
+    {
+        // The helper needs to be used to read contract state because
+        // AdminUpgradeabilityProxy does not allow the proxy admin to make
+        // proxy calls.
+
+        // Check that this contract sufficient funds to run the tests
+        uint256 contractBal = _helper.balanceOf(address(this));
```

Comment:
should we validate that `_helper` is using the same proxy address as the one we're upgrading?

Reply:
Sure, we can do that.

Reply:
I modified it such that the upgrader contract deploy the helper contract on its own.